commons/zenoh-shm/src/posix_shm/segment_lock.rs

```diff
@@ -0,0 +1,135 @@
+    use zenoh_result::ZResult;
+
+    #[repr(transparent)]
+    pub(crate) struct ShmLock(LockInner);
+
+    impl Drop for ShmLock {
+        fn drop(&mut self) {
+            if self.0._tempfile.try_lock(FileLockMode::Exclusive).is_ok() {
+                let _ = std::fs::remove_file(self.0.path.clone());
+            }
+        }
+    }
+
+    impl ShmLock {
+        pub(crate) fn create<T>(os_id: &T) -> ZResult<Self>
+        where
+            T: ?Sized + AsRef<OsStr>,
+        {
+            // calculate tempfile path
+            let path = tmp_file_path(os_id);
+
+            // create tempfile just to lock it
+            let tempfile = OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(path.clone())?;
+
+            // lock tempfile with shared lock to indicate that file is managed
+            tempfile.try_lock(FileLockMode::Shared)?;
+
+            Ok(Self(LockInner {
+                path,
+                _tempfile: tempfile,
+            }))
+        }
+
+        pub(crate) fn open<T>(os_id: &T) -> ZResult<Self>
+        where
+            T: ?Sized + AsRef<OsStr>,
+        {
+            // calculate tempfile path
+            let path = tmp_file_path(os_id);
+
+            // open tempfile just to lock it
+            let tempfile = OpenOptions::new().read(true).open(path.clone())?;
+
+            // lock tempfile with shared lock to indicate that file is managed
+            tempfile.try_lock(FileLockMode::Shared)?;
+
+            Ok(Self(LockInner {
+                path,
+                _tempfile: tempfile,
+            }))
+        }
+    }
+
+    #[repr(transparent)]
+    pub(crate) struct ExclusiveShmLock(LockInner);
+
+    impl ExclusiveShmLock {
+        pub(crate) fn open_exclusive<T>(os_id: &T) -> ZResult<Self>
+        where
+            T: ?Sized + AsRef<OsStr>,
+        {
+            // calculate tempfile path
+            let path = tmp_file_path(os_id);
+
+            // create or open tempfile just to lock it
+            let tempfile = OpenOptions::new()
+                .write(true)
+                .truncate(false)
+                .create(true)
+                .open(path.clone())?;
+
+            // lock tempfile with exclusive lock to guarantee that the file is unmanaged
+            tempfile.try_lock(FileLockMode::Exclusive)?;
+
+            Ok(Self(LockInner {
+                path,
+                _tempfile: tempfile,
+            }))
+        }
+    }
+
+    impl Drop for ExclusiveShmLock {
+        fn drop(&mut self) {
+            let _ = std::fs::remove_file(self.0.path.clone());
+        }
+    }
+
+    impl TryFrom<ShmLock> for ExclusiveShmLock {
+        type Error = ();
+
+        fn try_from(value: ShmLock) -> Result<Self, Self::Error> {
+            if value.0._tempfile.try_lock(FileLockMode::Exclusive).is_ok() {
+                return Ok(unsafe { core::mem::transmute::<ShmLock, ExclusiveShmLock>(value) });
```

Comment:
Technically this doesn't need `unsafe` as it's possible to write `ExclusiveShmLock(value.0)`. This would also avoid the need for `#[repr(transparent)]`.

Reply:
Compiler error:
"cannot move out of type `posix_shm::segment_lock::unix::ShmLock`, which implements the `Drop` trait"
prevents from doing this.
ShmLock need to have custom Drop trait implementation

Reply:
I understand!